src/go/rpk/pkg/cli/topic/trim.go

```diff
@@ -0,0 +1,279 @@
+		Partition int32  `yaml:"partition" json:"partition"`
+		Offset    int64  `yaml:"offset" json:"offset"`
+	}](fs, file)
+	if err == nil {
+		o := make(kadm.Offsets)
+		for _, element := range parsed {
+			o.Add(kadm.Offset{
+				Topic:     element.Topic,
+				Partition: element.Partition,
+				At:        element.Offset,
+			})
+		}
+		return o, nil
+	}
+	return nil, err
+}
+
+// parseOffsetFlag parse the offset string that we get from the flag, it parses
+// a simple integer or a timestamp-based offset (prefixed with @) it follows the
+// same syntax/decode as 'rpk topic consume --offset' flag.
+func parseOffsetFlag(ctx context.Context, adm *kadm.Client, topic, offset string, partition int) (int64, error) {
+	// Timestamp offset
+	if strings.HasPrefix(offset, "@") {
+		offset = offset[1:] // strip @
+		now := time.Now()
+		_, startAt, end, _, err := parseTimestampBasedOffset(offset, time.Now())
+		if err != nil {
+			return 0, err
+		} else if end {
+			return 0, errors.New("invalid offset '@end'; use '--to-hwm' if you want to trim all")
+		} else if now.Before(startAt) {
+			return 0, fmt.Errorf("invalid offset %q; rpk topic trim does not support a time onward,  use '--to-hwm' if you want to trim all", offset)
+		}
+
+		listed, err := adm.ListOffsetsAfterMilli(ctx, startAt.UnixMilli(), topic)
+		if err == nil {
+			err = listed.Error()
+		}
+		if err != nil {
+			return 0, fmt.Errorf("unable to list offsets after milli %d: %v", startAt.UnixMilli(), err)
+		}
+		filteredOffset, ok := listed.Lookup(topic, int32(partition))
+		if !ok {
+			return 0, fmt.Errorf("unable to find the offset after milli %d in the topic %q and partition %v", startAt.UnixMilli(), topic, partition)
+		}
+		return filteredOffset.Offset, nil
+	}
+	// Not @-prefixed offset
+	return strconv.ParseInt(offset, 10, 64)
+}
+
+// printDeleteRecordRequest prints the passed kadm.Offsets (used for the
+// deleteRecords request). It finds the start offset so the user can see what's
+// the change that is being done, it also finds the hwm for when the user passes
+// the '--to-hwm' flag.
+func printDeleteRecordRequest(ctx context.Context, adm *kadm.Client, o kadm.Offsets) (rerr error) {
+	tw := out.NewTable("TOPIC", "PARTITION", "PRIOR-START-OFFSET", "NEW-START-OFFSET")
+	o.Each(func(offset kadm.Offset) {
+		listed, err := adm.ListStartOffsets(ctx, offset.Topic)
+		if err == nil {
+			err = listed.Error()
+		}
+		if err != nil {
+			rerr = fmt.Errorf("unable to list start offset for topic %q: %v", offset.Topic, err)
+			return
+		}
+		l, ok := listed.Lookup(offset.Topic, offset.Partition)
+		if !ok {
+			rerr = fmt.Errorf("unable to find the start offset for the topic %q and partition %v", offset.Topic, offset.Partition)
+			return
+		}
+		// -1 maps to the partition's high watermark, and we want to print that
+		// instead of the '-1' that might be confusing for the end user.
+		if offset.At == -1 {
+			end, err := adm.ListEndOffsets(ctx, offset.Topic)
```

Comment:
Same thought here -- this needs to be done only once.

Reply:
https://github.com/redpanda-data/redpanda/pull/11825#discussion_r1252237254

Reply:
See my reply above, but more importantly: I don't think this branch makes sense, actually. -1 is only used for _input_ -- it is not returned from ListStartOffsets unless there is literally no offset that exists for this partition. This branch can be removed.

Reply:
I'll move the listEndOffset above.

> I don't think this branch makes sense, actually. -1 is only used for input

Correct. -1 is used for input to the DeleteRecords request. (L150 we set the `-1`)

> it is not returned from ListStartOffsets unless there is literally no offset that exists for this partition.

We are not evaluating what comes from ListStartOffsets here, we are evaluating if `offset.At` which comes from the request.

> This branch can be removed.

I don't think it's right to remove this unless I'm not seeing something. In Line 150 if the `--offset` flag is `--offset end` we set the kadm.Offset to be "-1" so the request to DeleteRecords will set the new low watermark to be the high watermark.

In this branch/conditional we are trying to print the request, this conditional is only to avoid displaying the new start offset as `-1`:

```
$rpk topic trim-prefix test --offset end 
TOPIC  PARTITION  PRIOR-START-OFFSET  NEW-START-OFFSET
test   0          12                  -1
```

And instead, show:

```
$rpk topic trim-prefix test --offset end 
TOPIC  PARTITION  PRIOR-START-OFFSET  NEW-START-OFFSET
test   0          12                  50
```